### PR TITLE
Feature/10349 crypto functions part1/10437 generating an ec key pair with curve p 256

### PR DESCRIPTION
**Pull request by 30mar**
## Description
Add ec key pair generation

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10437

The screenshot is of a command line check to check if the generated public key string is valid
## Screenshots
Screenshot 2021-11-15 at 15 36 14

### CODE REVIEW

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        return (modifiedData, nil)
+    }
+    
+    // MARK: - Private
+    
+    private func tag(for name: String) -> Data {
+      "\(Bundle.main.bundleIdentifier ?? "app").\(name)".data(using: .utf8)!
+    }
+    
+    private func appendHeaderToData(data: Data) -> Data {
```
Comment (marcussc) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift:71]:
```suggestion
    private func prependHeaderToData(data: Data) -> Data {
```

File: src/xcode/ENA/ENASecurity/Tests/ENASecurityTests/ECKeyPairGenerationTests.swift [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        let generationClass = ECKeyPairGeneration()
+        let generationOutput = generationClass.generatePrivateKey(with: "test")
+        guard let privateKey = generationOutput.0,
+              let publicKey = generationClass.generatePublicKey(from: privateKey) else {
+                  XCTFail()
+                  return
+              }
+        
+        guard let publicKeyData = generationClass.generateData(from: publicKey).0 else {
+            XCTFail()
```
Comment (marcussc) [inline src/xcode/ENA/ENASecurity/Tests/ENASecurityTests/ECKeyPairGenerationTests.swift:22]: An Error message would be nice so we can see directly on the CI what went wrong here.

File: src/xcode/ENA/ENASecurity/Tests/ENASecurityTests/ECKeyPairGenerationTests.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+@testable import ENASecurity
+
+final class ECKeyPairGenerationTests: XCTestCase {
+    
+    func test_validPublicKeyBase64_isGenerated() {
+        let generationClass = ECKeyPairGeneration()
+        let generationOutput = generationClass.generatePrivateKey(with: "test")
+        guard let privateKey = generationOutput.0,
+              let publicKey = generationClass.generatePublicKey(from: privateKey) else {
+                  XCTFail()
```
Comment (marcussc) [inline src/xcode/ENA/ENASecurity/Tests/ENASecurityTests/ECKeyPairGenerationTests.swift:17]: And Error message would be nice so we can see directly on the CI what went wrong here.

**Review by marcussc — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+
+public struct ECKeyPairGeneration {
+    
+    // MARK: - Public
+
+    public func generatePrivateKey(with name: String? = nil) -> (SecKey?, String?) {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift:11]: I think we should have a common API Style. In the other cases we are using Result type with explicit errors.

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+
+public struct ECKeyPairGeneration {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift:7]: API which returns 3 values, like described in the tech spec, is missing.

**Review by ArturFriesen — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    case dataGenerationFromKeyFailed(String?) // if key cannot be used for AES
+}
+
+public struct ECKeyPairGeneration {
+    
+    // MARK: - Public
+
+    func generateECPair(with name: String? = nil) -> Result<ECKeyPair, ECKeyPairGenerationError> {
+        let secureKey = self.generatePrivateKey(with: name)
+        guard let privateKey = secureKey.0 else {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift:19]: Not mandatory, just for information: If the other functions would also return a Result type, you could just concatenate all the calls with flatmap. This way this func will only be a couple of lines. (https://useyourloaf.com/blog/using-swift-result-and-flatmap/)

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift (hunk 6 above)
Comment (30mar) [reply src/xcode/ENA/ENASecurity/Sources/ENASecurity/ECKeyPairGeneration.swift:19]: they are not returning a Result type, they return an error message or no error and based on the function i return the needed enum

**Review by marcussc — APPROVED** [review]

**Pull request merged**